File: typescript/src/types/user-groups.ts
import { DateTime, UUID } from './common';

/**
 * User subgroup response object
 */
export interface UserSubgroup {
  id: UUID;
  groupId: UUID;
  name: string;
  slug: string;
  description: string | null;
  sortOrder: number;
  active: boolean;
  userCount?: number;
  createdAt: DateTime;
  updatedAt: DateTime;
}

/**
 * User group response object
 */
export interface UserGroup {
  id: UUID;
  name: string;
  slug: string;
  description: string | null;
  color: string;
  sortOrder: number;
  active: boolean;
  userCount: number;
  subgroups: UserSubgroup[];
  createdAt: DateTime;
  updatedAt: DateTime;
}

/**
 * User group tree response (compact)
 */
export interface UserGroupTree {
  id: UUID;
  name: string;
  slug: string;
  color: string;
  userCount: number;
  subgroups: Array<{
    id: UUID;
    name: string;
    slug: string;
  }>;
}

/**
 * Create user group request body
 */
export interface CreateUserGroupBody {
  name: string;
  slug?: string;
  description?: string | null;
  color?: string;
  sortOrder?: number;
  active?: boolean;
}

/**
 * Update user group request body
 */
export interface UpdateUserGroupBody {
  name?: string;
  slug?: string;
  description?: string | null;
  color?: string;
  sortOrder?: number;
  active?: boolean;
}

/**
 * Create user subgroup request body
 */
export interface CreateUserSubgroupBody {
  name: string;
  slug?: string;
  description?: string | null;
  sortOrder?: number;
  active?: boolean;
}

/**
 * Update user subgroup request body
 */
export interface UpdateUserSubgroupBody {
  name?: string;
  slug?: string;
  description?: string | null;
  sortOrder?: number;
  active?: boolean;
}

/**
 * Assign user to group request body
 */
export interface AssignUserToGroupBody {
  userId: UUID;
  groupId?: UUID | null;
  subgroupId?: UUID | null;
}

/**
 * Bulk assign users request body
 */
export interface BulkAssignUsersBody {
  userIds: UUID[];
  groupId?: UUID | null;
  subgroupId?: UUID | null;
}

/**
 * User group list query parameters
 */
export interface UserGroupListParams {
  active?: boolean;
}

/**
 * User group statistics response
 */
export interface UserGroupStats {
  total: number;
  active: number;
  inactive: number;
  totalUsers: number;
}

File: python/sakha_cms/models/user_groups.py
"""User Groups module models."""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from typing import Optional


class UserGroup(BaseModel):
    """User group."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    project_id: Optional[str] = Field(None, alias="project_id")
    name: str
    slug: str
    description: Optional[str] = None
    member_count: int = Field(default=0, alias="member_count")
    is_active: bool = Field(default=True, alias="is_active")
    permissions: list[str] = Field(default_factory=list)
    created_at: datetime = Field(alias="created_at")
    updated_at: datetime = Field(alias="updated_at")


class CreateUserGroupBody(BaseModel):
    """Create user group request."""

    model_config = ConfigDict(populate_by_name=True)

    name: str
    slug: Optional[str] = None
    description: Optional[str] = None
    permissions: Optional[list[str]] = None


class UpdateUserGroupBody(BaseModel):
    """Update user group request."""

    model_config = ConfigDict(populate_by_name=True)

    name: Optional[str] = None
    slug: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = Field(None, alias="isActive")
    permissions: Optional[list[str]] = None


class UserGroupStats(BaseModel):
    """User group statistics."""

    model_config = ConfigDict(populate_by_name=True)

    total_groups: int = Field(alias="totalGroups")
    active_groups: int = Field(alias="activeGroups")
    total_members: int = Field(alias="totalMembers")
    average_group_size: float = Field(alias="averageGroupSize")
